src/dune_pkg/package_variable.ml

```diff
@@ -0,0 +1,77 @@
+open Stdune
+open Dune_lang
+
+module Name = struct
+  include String
+
+  include (
+    Dune_util.Stringlike.Make (struct
+      type t = string
+
+      let to_string x = x
+      let module_ = "Package_variable.Name"
+      let description = "package variable name"
+      let description_of_valid_string = None
+      let hint_valid = None
+      let of_string_opt s = if s = "" then None else Some s
+    end) :
+      Dune_util.Stringlike with type t := t)
+end
+
+module Scope = struct
+  type t =
+    | Self
+    | Package of Package_name.t
```

Comment:
FTR OPAM also knows of `Global`. I assume we can ignore those since this is only for package variables.

But on the other hand, should we have a case for `Self`? It is a more specific case of `Package` with the package name matching its own package. Maybe these should be normalized to regular package variables (that is, fill in `_` with the package name, as it is done already in the `macro` expansion, but earlier).

What do you think?

Reply:
> Maybe these should be normalized to regular package variables (that is, fill in _ with the package name, as it is done already in the macro expansion, but earlier).

That would create unnecessary noise when renaming a package.

Reply:
How so? If a package is renamed, we generate a new lock file with a new name anyway and the names in these files are generated by us in any case.

Reply:
Without this PR, git would figure out that all you've done is move `x.pkg` to `y.pkg`. With this PR, regenerating the lock file would now replace every instance of `x` by `y` inside `y.pkg` which would trick git that this file was written from scratch.

Reply:
Fair enough, however how often do renames happen? I can't remember any package being renamed, because it breaks all its dependencies; the only case of that happening is when OUnit2 renamed itself using an intermediate package.

Reply:
Didn't the lock free data structures library get recently renamed to saturn? https://github.com/ocaml-multicore/saturn

It's not very common for packages to be renamed but also not unheard of.

Reply:
I think `kcas` was joined with a few other packages into `saturn`, not a straight-up rename. However, I am not very involved what happened, so I might have gotten it wrong.

My actual point is that maybe the change is still worth it for a simpler semantic in Dune by just having one kind of package variables instead of optimizing for renames of packages.

Reply:
I do think there are solutions that are about as simple that don't interfere with renaming. In my other comment, I mentioned introducing a different form for self packages for example.